File: solutions/python/diffie-hellman/3/diffie_hellman.py
"""
This module provides functions for generating private and public Diffie-Hellman keys,
as well as computing the shared secret key.
"""

import secrets


def private_key(prime_modulus):
    """
    Generates a private key for use in Diffie-Hellman key exchange.

    Args:
        prime_modulus (int): The prime modulus p.

    Returns:
        int: A randomly chosen private key.
    """
    return secrets.choice(range(2, prime_modulus))


def public_key(prime_modulus, generator, private_key):
    """
    Computes a public key for use in Diffie-Hellman key exchange.

    Args:
        prime_modulus (int): The prime modulus p.
        generator (int): The base generator g.
        private_key (int): The private key.

    Returns:
        int: The computed public key.
    """
    return (generator**private_key) % prime_modulus


def secret(prime_modulus, public_key, private_key):
    """
    Computes the shared secret using the private key and the peer's public key.

    Args:
        prime_modulus (int): The prime modulus p.
        public_key (int): The peer's public key.
        private_key (int): The private key.

    Returns:
        int: The computed shared secret.
    """
    return (public_key**private_key) % prime_modulus

File: solutions/javascript/diffie-hellman/2/diffie-hellman.js
/**
 * Check if a number is a prime number.
 *
 * A prime number is a number greater than 1 that is only divisible by 1 and itself.
 *
 * @param {number} number - The number to test for primality.
 * @returns {boolean} Returns `true` if the number is prime, otherwise `false`.
 */
const isPrime = (number) => {
  if (number <= 1) {
    return false;
  }
  if (number === 2 || number === 3) {
    return true;
  }
  if (number % 2 === 0) {
    return false;
  }
  for (let i = 3; i <= Math.sqrt(number); i += 2) {
    if (number % i === 0) {
      return false;
    }
  }
  return true;
};

/**
 * Class representing the Diffie-Hellman key exchange protocol.
 *
 * This class provides methods to generate public keys and shared secrets
 * using the Diffie-Hellman algorithm. The class assumes integer arithmetic.
 */
export class DiffieHellman {
  /**
   * Create a Diffie-Hellman instance.
   *
   * @param {number} p - A large prime number (modulus).
   * @param {number} g - A prime number (base/generator).
   * @throws {Error} Throws an error if either `p` or `g` is not a prime number
   *                 or falls outside the valid range for Diffie-Hellman.
   */
  constructor(p, g) {
    if (!DiffieHellman.validateInitialArguments(p, g)) {
      throw Error('Constructor arguments are out of range or non-prime!');
    }

    this.p = p;
    this.g = g;
  }

  /**
   * Generate the public key using a private key.
   *
   * @param {number} privateKey - A private key (must be > 1 and < p).
   * @returns {number} The corresponding public key.
   * @throws {Error} Throws an error if the private key is ≤ 1 or ≥ p.
   */
  getPublicKey(privateKey) {
    if (privateKey <= 1 || privateKey >= this.p) {
      throw Error(
        'Private key must be greater than 1 and less than the modulus parameter p!',
      );
    }

    return this.g ** privateKey % this.p;
  }

  /**
   * Generate a random private key for a given prime modulus.
   *
   * @param {number} p - The prime modulus.
   * @returns {number} A random private key in the range (1, p).
   */
  static getPrivateKey(p) {
    return Math.floor(Math.random() * (p - 2) + 2);
  }

  /**
   * Compute the shared secret using another party's public key and our private key.
   *
   * @param {number} theirPublicKey - The other party's public key.
   * @param {number} ourPrivateKey - Our own private key.
   * @returns {number} The shared secret value.
   */
  getSecret(theirPublicKey, ourPrivateKey) {
    return theirPublicKey ** ourPrivateKey % this.p;
  }

  /**
   * Validate the initial prime parameters `p` and `g`.
   *
   * @param {number} p - A prime number (modulus).
   * @param {number} g - A prime number (generator).
   * @returns {boolean} Returns `true` if both values are prime, otherwise `false`.
   */
  static validateInitialArguments(p, g) {
    return isPrime(p) && isPrime(g);
  }
}
